airbyte-integrations/connectors/source-oracle-v2/src/cdk/main/kotlin/io/airbyte/cdk/command/ConfigurationJsonObjectSupplier.kt

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.command
+
+import com.fasterxml.jackson.databind.JsonNode
+import io.micronaut.context.annotation.Value
+import jakarta.inject.Singleton
+import java.util.function.Supplier
+
+/**
+ * Supplies a valid [T] configuration POJO instance, based on the `airbyte.connector.config`
+ * Micronaut property values:
+ * - either `airbyte.connector.config.json` if it is set (typically by the CLI)
+ * - or the other, nested `airbyte.connector.config.*` properties (typically in unit tests)
+ *
+ * One may wonder why we need to inject this [Supplier] instead of injecting the POJO directly. The
+ * reason is that injecting the POJO only works if the configuration values are set via the nested
+ * Micronaut properties (i.e. in unit tests). We could make direct injection work the same way as
+ * the [ConfiguredCatalogFactory] or the [InputStateFactory] (via a @Factory) but then we'd lose the
+ * ability to set values via the nested properties. This current design caters to both use cases.
+ * Furthermore, by deferring the parsing and validation of the configuration, we don't need to worry
+ * about exception handling edge cases when implementing the CHECK operation.
+ *
+ * The object is also validated against its [jsonSchema] JSON schema, derived from [javaClass].
+ */
+@Singleton
+class ConfigurationJsonObjectSupplier<T : ConfigurationJsonObjectBase>(
+    private val micronautPropertiesFallback: T,
+    @Value("\${${CONNECTOR_CONFIG_PREFIX}.json}") private val jsonPropertyValue: String? = null
+) : Supplier<T> {
+
+    @Suppress("UNCHECKED_CAST")
+    val javaClass: Class<T> = micronautPropertiesFallback::class.java as Class<T>
+
+    val jsonSchema: JsonNode by lazy { JsonUtils.generator.generateJsonSchema(javaClass) }
+
+    override fun get(): T {
+        val json: String =
+            jsonPropertyValue
+                ?: JsonUtils.jsonNode(javaClass, micronautPropertiesFallback).toString()
+        return JsonUtils.parseOne(javaClass, json)
+    }
+}
+
+/**
+ * Connector configuration POJO supertype.
+ *
+ * This dummy base class is required by Micronaut. Without it, thanks to Java's type erasure, it
+ * thinks that the [ConfigurationJsonObjectSupplier] requires a constructor argument of type [Any].
+ *
+ * Strictly speaking, its subclasses are not really POJOs anymore, but who cares.
+ */
+abstract class ConfigurationJsonObjectBase
```

Comment:
I think it's totally fine to have a base class. I would just rename it into BaseConfiguration or AbstractConfiguration. I think the Json aspect is unnecessary, as is the fact that it's an Object

Reply:
The name is to signify that it's the configuration in the json object model, not the internal connector data model (implementations of `Configuration` and `SourceConfiguration`)